milestones/data.py

```diff
@@ -0,0 +1,373 @@
+        milestone=milestone.id).delete()
+    internal.UserMilestone.objects.filter(
+        milestone=milestone.id).delete()
+    internal.Milestone.objects.filter(
+        id=milestone.id).delete()
+
+
+def fetch_milestones(milestone):
+    """
+    Retrieves a set of matching milestones from app/local state
+    Returns a list-of-dicts representation of the object
+    """
+    if milestone is None:
+        raise exceptions.InvalidMilestoneException()
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    if milestone_obj.id is not None:
+        return serializers.serialize_milestones(internal.Milestone.objects.filter(
+            id=milestone_obj.id,
+            active=True,
+        ))
+    if milestone_obj.namespace is not None:
+        return serializers.serialize_milestones(internal.Milestone.objects.filter(
+            namespace=milestone_obj.namespace,
+            active=True
+        ))
+    return []
+
+
+def create_course_milestone(course_key, relationship, milestone):
+    """
+    Inserts a new course-milestone into app/local state
+    No response currently defined for this operation
+    """
+    relationship_type = _get_milestone_relationship_type(relationship)
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    internal.CourseMilestone.objects.get_or_create(
+        course_id=unicode(course_key),
+        milestone=milestone_obj,
+        milestone_relationship_type=relationship_type,
+        active=True,
+    )
+
+
+def delete_course_milestone(course_key, milestone):
+    """
+    Removes an existing course-milestone from app/local state
+    No response currently defined for this operation
+    """
+    try:
+        internal.CourseMilestone.objects.get(
+            course_id=unicode(course_key),
+            milestone=milestone['id'],
+            active=True,
+        ).delete()
+    except internal.CourseMilestone.DoesNotExist:
+        pass
+
+
+def fetch_courses_milestones(course_keys, relationship=None, user=None):
+    """
+    Retrieves the set of milestones currently linked to the specified courses
+    Optionally pass in 'relationship' (ex. 'fulfills') to filter down the set
+    """
+    queryset = internal.CourseMilestone.objects.filter(
+        course_id__in=course_keys,
+        active=True
+    ).select_related('milestone')
+
+    # if milestones relationship type found then apply the filter
+    if relationship is not None:
+        mrt = _get_milestone_relationship_type(relationship)
+        queryset = queryset.filter(
+            milestone_relationship_type=mrt.id,
+        )
+
+    # To pull the list of milestones a user HAS, use get_user_milestones
+    # Use fetch_courses_milestones to pull the list of milestones that a user does not yet
+    # have for the specified course
+    if relationship == 'requires' and user and user.get('id', 0) > 0:
+        queryset = queryset.exclude(milestone__usermilestone__user_id=user['id'])
```

Comment:
We need to apply the user filtering in this case in order to see which milestones are required by a course that a user has not already collected.   At present I didn't see a need to support filtering by user for the 'fulfills' relationship, and wanted to avoid applying a filter (and doing the ensuing work) if a client attempted to filter on this combo.

Maybe it makes more sense to throw an exception for scenarios not involving 'requires' and a user filter, and future developers can modify as needed?